packages/gatsby-source-shopify/src/clients.ts

```diff
@@ -0,0 +1,95 @@
+import fetch, { Response } from "node-fetch"
+import { HttpError } from "./errors"
+
+const MAX_BACKOFF_MILLISECONDS = 60000
+
+interface IGraphQLClient {
+  request: <T>(query: string, variables?: Record<string, any>) => Promise<T>
+}
+
+interface IRestClient {
+  request: (path: string) => Promise<Response>
+}
+
+export function createGraphqlClient(
+  options: IShopifyPluginOptions
+): IGraphQLClient {
+  const url = `https://${options.storeUrl}/admin/api/2021-07/graphql.json`
+
+  async function graphqlFetch<T>(
+    query: string,
+    variables?: Record<string, any>,
+    retries = 0
+  ): Promise<T> {
+    const response = await fetch(url, {
+      method: `POST`,
+      headers: {
+        "Content-Type": `application/json`,
+        "X-Shopify-Access-Token": options.password,
+      },
+      body: JSON.stringify({
+        query,
+        variables,
+      }),
+    })
+
+    if (!response.ok) {
+      const waitTime = 2 ** (retries + 1) + 500
+      if (response.status >= 500 && waitTime < MAX_BACKOFF_MILLISECONDS) {
+        await new Promise(resolve => setTimeout(resolve, waitTime))
+        return graphqlFetch(query, variables, retries + 1)
+      }
+
+      throw new HttpError(response)
+    }
+
+    const json = await response.json()
+    return json.data as T
+  }
+
+  return { request: graphqlFetch }
+}
+
+export function createRestClient(options: IShopifyPluginOptions): IRestClient {
+  const baseUrl = `https://${options.storeUrl}/admin/api/2021-01`
+
+  async function shopifyFetch(
+    path: string,
+    fetchOptions = {
+      headers: {
+        "X-Shopify-Access-Token": options.password,
+      },
+    },
+    retries = 3
+  ): Promise<Response> {
+    /**
+     * This is kind of a hack, but...
+     *
+     * We do this because although callers will use a relative path,
+     * some responses might have pagination links that get fed back
+     * to shopifyFetch to retrieve the next page. Those links are
+     * absolute URLs so we account for both, but not in a very robust
+     * fashion.
+     */
+    const url = path.includes(options.storeUrl) ? path : `${baseUrl}${path}`
+
+    const resp = await fetch(url, fetchOptions)
+
+    if (!resp.ok) {
```

Comment:
```suggestion
    if (!resp.ok && retries > 0 && resp.status === 429) {
```

Reply:
do we want to only refetch on rate limit errors? It could be some 500 error as well that gets resolved as well.

Reply:
This is actually a bit of the code I chose not to change. I just pulled the rest and graphql clients into a single file, but I didn't change the code itself. I'm more than happy to do some refactoring if you guys see the need.

Reply:
Plus one on 500 errors. Can you add a check for those too?